zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+                    maybe_pool.as_mut().ok_or::<DispatchError>(Error::<T>::PoolNotFound.into())?;
+                ensure!(
+                    pool.liquidity_shares_manager.fees == Zero::zero(),
+                    Error::<T>::OutstandingFees
+                );
+                let ratio = bdiv(
+                    pool_shares_amount.saturated_into(),
+                    pool.liquidity_shares_manager.total_shares()?.saturated_into(),
+                )?;
+                let mut amounts_out = vec![];
+                for (&asset, &min_amount_out) in pool.assets().iter().zip(min_amounts_out.iter()) {
+                    let balance_in_pool = pool.reserve_of(&asset)?;
+                    let amount_out: BalanceOf<T> =
+                        bmul(ratio, balance_in_pool.saturated_into())?.saturated_into();
+                    amounts_out.push(amount_out);
+                    ensure!(amount_out >= min_amount_out, Error::<T>::AmountOutBelowMin);
+                    T::AssetManager::transfer(asset, &pool.account_id, &who, amount_out)?;
+                }
+                for ((_, balance), &amount_out) in pool.reserves.iter_mut().zip(amounts_out.iter())
+                {
+                    *balance = balance.saturating_sub(amount_out);
+                }
+                pool.liquidity_shares_manager.exit(&who, pool_shares_amount)?;
+                if pool.liquidity_shares_manager.total_shares()? == Zero::zero() {
+                    // FIXME We will withdraw all remaining funds (there shouldn't be any unless
+                    // we're using the native currency). The implementation is a bit defensive here.
+                    // This is an ugly hack and system should offer the option to whitelist
+                    // accounts.
+                    if pool.collateral == Asset::Ztg {
+                        let remaining = T::AssetManager::free_balance(Asset::Ztg, &pool.account_id);
+                        T::AssetManager::withdraw(Asset::Ztg, &pool.account_id, remaining)?;
+                    }
+                    *maybe_pool = None; // Delete the storage map entry.
+                    Self::deposit_event(Event::<T>::PoolDestroyed {
+                        who: who.clone(),
+                        market_id,
+                        pool_shares_amount,
+                        amounts_out,
+                    });
+                } else {
+                    let liq = pool.liquidity_parameter;
+                    let new_liquidity_parameter = liq.saturating_sub(
+                        bmul(ratio.saturated_into(), liq.saturated_into())?.saturated_into(),
+                    );
+                    ensure!(
+                        new_liquidity_parameter >= MIN_LIQUIDITY.saturated_into(),
+                        Error::<T>::LiquidityTooLow
+                    );
+                    pool.liquidity_parameter = new_liquidity_parameter;
+                    Pools::<T>::insert(market_id, pool);
```

Comment:
Isn't this already covered by the surrounding `try_mutate_pool` call?

```suggestion
```

Reply:
9e02d068990fe8baf5f1b0525fff7188d363619d